testing/helpers/matrix-test-helper.js

```diff
@@ -1,33 +1,255 @@
+  const wrap = (x) => RegExp(x || '.*', 'i');
+  return {
+    product: wrap(product),
+    demo: wrap(demo),
+    framework: wrap(framework),
+  };
+}
+
+function getInterestProcessArgs() {
+  // eslint-disable-next-line spellcheck/spell-checker
+  return process.argv.slice(2);
+}
+
+export function getExplicitTestsFromArgs() {
+  const result = { masks: [] };
+  getInterestProcessArgs().forEach((argument) => {
+    const parts = argument.split('-');
+    result.masks.push(patternGroupFromValues(...parts));
+  });
+  return result.masks.length ? result : undefined;
+}
+
+export function getChangedFiles() {
+  const changedFilesPath = process.env.CHANGEDFILEINFOSPATH;
+  return changedFilesPath
+    && existsSync(changedFilesPath)
+    && JSON.parse(readFileSync(changedFilesPath));
+}
+
+function getExplicitTestsInternal() {
+  const changedFiles = getChangedFiles();
+
+  if (!changedFiles) { return getExplicitTestsFromArgs(); }
+
+  const result = { masks: [], traceTree: undefined };
+
+  // eslint-disable-next-line no-restricted-syntax
+  for (const changedFile of changedFiles) {
+    const fileName = changedFile.filename;
+
+    if (settings.ignoreChangesPathPatterns.some((x) => x.test(fileName))) {
+      // eslint-disable-next-line no-continue
+      continue;
+    }
+
+    let parseResult = settings.demoExpr.exec(fileName)
+      || settings.demoFilesExpr.exec(fileName)
+      || settings.commonEtalonsExpr.exec(fileName);
+    // eslint-disable-next-line no-cond-assign
+    if (!parseResult && (parseResult = settings.manualEtalonsExpr.exec(fileName))) {
+      if (result.traceTree == null) { result.traceTree = {}; }
+      let currentNamePart = result.traceTree;
+      // eslint-disable-next-line no-restricted-syntax
+      for (const fileNamePart of fileName.split(/[/\\]/)) {
+        const nextNamePart = currentNamePart[fileNamePart] || {};
+        currentNamePart[fileNamePart] = nextNamePart;
+        currentNamePart = nextNamePart;
+      }
+      // eslint-disable-next-line no-continue
+      continue;
+    }
+
+    if (parseResult) {
+      parseResult.groups = parseResult.groups || {};
+      if (parseResult.groups.data && parseResult.groups.data === 'description.md') {
+        // eslint-disable-next-line no-continue
+        continue;
+      }
+
+      result.masks.push(patternGroupFromValues(
+        parseResult.groups.product,
+        parseResult.groups.demo,
+        parseResult.groups.framework,
+      ));
+    } else {
+      // eslint-disable-next-line no-console
+      if (settings.verbose) { console.log('Unable to parse changed file, running all tests: ', fileName); }
+      return undefined;
     }
   }
+
+  return result;
+}
+function getExplicitTests() {
+  const result = getExplicitTestsInternal();
+  if (result) {
+    const oldToJSON = RegExp.prototype.toJSON;
+    try {
+      // eslint-disable-next-line no-extend-native
+      RegExp.prototype.toJSON = RegExp.prototype.toString;
```

Comment:
Why is this magic necessary?

Reply:
https://stackoverflow.com/questions/12075927/serialization-of-regexp